Create a navigation delegate to surface lookalike character sanitization on navigation

charliewolfe opened the pull request:
#### 714e9f81cb312f7dc36bd48ef7bb0fcb2f320f9f
<pre>
Create a navigation delegate to surface lookalike character sanitization on navigation
<a href="https://bugs.webkit.org/show_bug.cgi?id=250104">https://bugs.webkit.org/show_bug.cgi?id=250104</a>
rdar://103888086

Reviewed by Wenson Hsieh.

This adds a navigation delegate method which will be used to
surface lookalike character changes on navigation.

* Source/WebKit/UIProcess/API/APINavigationClient.h:
(API::NavigationClient::didChangeLookalikeCharacters):
* Source/WebKit/UIProcess/API/Cocoa/WKNavigationDelegatePrivate.h:
* Source/WebKit/UIProcess/Cocoa/NavigationState.h:
* Source/WebKit/UIProcess/Cocoa/NavigationState.mm:
(WebKit::NavigationState::setNavigationDelegate):
(WebKit::NavigationState::NavigationClient::didChangeLookalikeCharacters):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didChangeLookalikeCharacters):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::sanitizeLookalikeCharacters):
(WebKit::WebPage::sanitizeLookalikeCharacters const): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: <a href="https://commits.webkit.org/258493@main">https://commits.webkit.org/258493@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e685940062c40fded8a9028db6d25529e1f185d8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102065 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11207 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/35131 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/111388 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/12180 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2118 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/94451 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/109128 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/107846 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/9327 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/92595 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/37146 "Passed tests") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/91207 "Found 1 new API test failure: TestWebKitAPI.WebKit.QuotaDelegateReload (failure)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/24077 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/78871 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/4769 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/25505 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/4876 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/1948 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/10935 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/44996 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5828 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/6625 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/UIProcess/WebPageProxy.messages.in:
```diff
@@ -628,4 +628,6 @@ messages -> WebPageProxy {
     ClassifyModalContainerControls(Vector<String> texts) -> (Vector<WebCore::ModalContainerControlType> types)
     DecidePolicyForModalContainer(OptionSet<WebCore::ModalContainerControlType> types) -> (enum:uint8_t WebCore::ModalContainerDecision decision)
+
+    DidChangeLookalikeCharacters(URL sourceURL, URL destinationURL)
```
Nit - maybe something like, `oldURL` / `newURL` or `originalURL` / `adjustedURL`?

whsieh commented on Source/WebKit/UIProcess/API/Cocoa/WKNavigationDelegatePrivate.h:
```diff
@@ -104,6 +104,7 @@ static const WKNavigationResponsePolicy _WKNavigationResponsePolicyBecomeDownloa
 - (void)_webView:(WKWebView *)webView didFinishLoadWithRequest:(NSURLRequest *)request inFrame:(WKFrameInfo *)frame WK_API_AVAILABLE(macos(11.0), ios(14.0));
 
 - (void)_webView:(WKWebView *)webView didFailLoadDueToNetworkConnectionIntegrityWithURL:(NSURL *)url WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA));
+- (void)_webView:(WKWebView *)webView didChangeLookalikeCharactersFromURL:(NSURL *)sourceURL toURL:(NSURL *)destinationURL WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA));
```
Nit - same here, I would probably call this `Original`/`Adjusted` or `Old`/`New` instead of source and destination.

charliewolfe replied:
I can make this change, but the reason I made it like this was to match `didPerformClientRedirectFromURL:(NSURL *)sourceURL toURL:(NSURL *)destinationURL` defined earlier in this file. Albeit the use for this is different, since it isn't actually a redirect in our case.

whsieh replied:
Ah, I see. I think it actually may be a tiny bit clearer to avoid using source/destination terminology here to avoid making it seem like the `sourceURL` is the currently loaded URL that we're navigating away from.

whsieh reviewed (approved):
Seems reasonable, aside from some minor naming suggestions.

charliewolfe committed:
Create a navigation delegate to surface lookalike character sanitization on navigation
https://bugs.webkit.org/show_bug.cgi?id=250104
rdar://103888086

Reviewed by Wenson Hsieh.

This adds a navigation delegate method which will be used to
surface lookalike character changes on navigation.

* Source/WebKit/UIProcess/API/APINavigationClient.h:
(API::NavigationClient::didChangeLookalikeCharacters):
* Source/WebKit/UIProcess/API/Cocoa/WKNavigationDelegatePrivate.h:
* Source/WebKit/UIProcess/Cocoa/NavigationState.h:
* Source/WebKit/UIProcess/Cocoa/NavigationState.mm:
(WebKit::NavigationState::setNavigationDelegate):
(WebKit::NavigationState::NavigationClient::didChangeLookalikeCharacters):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didChangeLookalikeCharacters):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::sanitizeLookalikeCharacters):
(WebKit::WebPage::sanitizeLookalikeCharacters const): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: https://commits.webkit.org/258493@main
